### PR TITLE
feat(ui): Implement 'attach --create' so attach would never fail

### PR DESCRIPTION
**Pull request by ottok**
Add support for `--create` in subcommand `attach` to enable similar use as in `tmux new-session -A -s Foo` where attaching a session will always result in a session, either re-using the old or creating a new.

This allows me to have a Zellij session open "forever", even across remote host reboots when the session otherwise dies. For example I can have `while ! ssh host -t 'zellij attach --create workspace'; do echo $?; sleep 3; done` and that window will forever be an open terminal to my remote machine with my previous session open as far as possible.

Closes: zellij-org/zellij#541

zellij-attach-create

### CODE REVIEW

**Comment by ottok** [comment]
PS. This is my first Rust project, so feel free to nitpick about all the things I am doing wrong so that I can learn.

**Comment by imsnif** [comment]
Hey @ottok - thanks for opening this PR, but unfortunately there's already one in the works for this: https://github.com/zellij-org/zellij/pull/587
I'd be very happy if you'd like to maybe pick up a different issue and work on it? We can make sure to mark that you're working on it so no one can take it away. Would also be happy to give it my personal attention and nitpick the code as much as you want :)

**Comment by ottok** [comment]
> First-time contributors need a maintainer to approve running workflows.

Could you please approve workflows so I can see if this PR would have passed tests even if I would decide to discard it?

**Comment by kunalmohan** [comment]
Hi @ottok! Looks like the author of #587 will be busy for some time (https://github.com/zellij-org/zellij/pull/587#issuecomment-877070373). So I'll review this and if this is ready before the other one, we can go ahead and merge this.

File: src/sessions.rs
Comment (kunalmohan) [inline src/sessions.rs:50]: This function looks redundant to me now. We can simply `unwrap()` on the result returned by `read_sessions()` to get the same behaviour.

File: src/main.rs
Comment (kunalmohan) [inline src/main.rs:102]: This should be checked only if there are no active sessions. If there is only one active session, we attach to it. If there are more than one active sessions we exit with a message. If there are no active sessions, we create one if `create` flag was passed, otherwise exit.

**Review by kunalmohan — CHANGES_REQUESTED** [review]
Hey @ottok! Thank you for the PR. This mostly looks good. There are a couple of changes that need to be done to handle all the cases correctly and a few other suggestions regarding the code. And make sure to fix the fmt and clippy lint errors. Feel free to ping me if you need any help 🙂

**Comment by ottok** [comment]
Apparently the workflows don't run when I refresh the PR. Which I did to make sure CI passes. But now I can't know if it does for sure..

File: src/main.rs
Comment (ottok) [reply src/main.rs:21]: But session_name might be empty. I'll think about something.

File: src/main.rs
Comment (ottok) [reply src/main.rs:99]: I have them at least now while developing, otherwise I don't get any feedback from Zellij on what code path it took.

File: src/main.rs
Comment (ottok) [reply src/main.rs:100]: Well this is how tmux does it to allow one to create a new fresh session if the previous one had died. I think this is useful.

If we wan't the `<name>` to be compulsory for `zellij attach --create <name>` then that should maybe be checked already in the command-line argument parsing phase.

Comment (kunalmohan) [reply src/main.rs:100]: I agree that we should allow users to create a fresh session if one does not exist. But this case here is the one where the user does not mention the session name. Since there is no session name, we will create a new session only if there are no active sessions. Otherwise, `zellij attach --create` will keep on creating a new session each time even though there are more than 1 active sessions and the user might want to attach to one of those.

IMO this is the behaviour that zellij should have with `--create` flag:
1. If session name is specified:
- If a session with name `<session_name>` is present, attach to it.
- If a session with name `<session_name>` does not exist, create a new session.

2. If no session name is specified:
-  If only a single session exists, attach to it.
-  If more than one session exists, exit with a message (and list the sessions).
-  If no session exists, create a new one with a random name

Do you have something else in mind?

File: src/main.rs
Comment (kunalmohan) [reply src/main.rs:21]: If the session_name is empty, the parameter `session_name` will be `None`. So, you can then operate on it the same way you do now:
```rust
fn start_new_session_name(name: Option<&String>) -> String {
    let session_name = name
        .map_or_else(|| names::Generator::default().next().unwrap(), |s| s.clone());
    assert_session_ne(&session_name);
    return session_name;
}
```

**Comment by kunalmohan** [comment]
Hi @ottok! Are you still working on this? This is an important feature and we would like to have it in main branch as soon as possible. In case you're busy, I'll be more than happy to take it off your hands :)

**Comment by ottok** [comment]
I will take a new look at it today. Did you have any follow-up comments to
my responses?

su 1. elok. 2021 klo 0.21 Kunal Mohan ***@***.***> kirjoitti:

**Comment by kunalmohan** [comment]
> Did you have any follow-up comments to my responses?

I had replied to them in the inline comments themselves. Are there more concerns?

File: src/main.rs
Comment (ottok) [reply src/main.rs:21]: I tried to get this working in various ways, but always every avenue I took ended up in failures. I guess I don't yet fully master the Rust ownership and types. Could you please show me what exactly I should have in the calling function?

Comment (kunalmohan) [reply src/main.rs:21]: Please see https://github.com/kunalmohan/zellij/commit/ec464a2f539b9fdf6d13e086420e034091c0fecb for the relevant changes. Feel free to reach out in case there's any confusion.

**Comment by a-kenji** [comment]
Thank you for the work on this feature!
It has now been implemented by #731.

**Comment by jeffschwartz** [comment]
Please update cli help to reflect new feature.

**Comment by ottok** [comment]
It is already in the cli:

```
± zellij attach --help
zellij-attach 
Attach to a session

USAGE:
    zellij attach [OPTIONS] [SESSION_NAME] [SUBCOMMAND]

ARGS:
    <SESSION_NAME>    Name of the session to attach to

OPTIONS:
    -c, --create                Create a session if one does not exist
    -f, --force-run-commands    If resurrecting a dead session, immediately run all its commands on
                                startup
    -h, --help                  Print help information
        --index <INDEX>         Number of the session index in the active sessions ordered creation
                                date

SUBCOMMANDS:
    help       Print this message or the help of the given subcommand(s)
    options    Change the behaviour of zellij
```

**Comment by jeffschwartz** [comment]
@ottok thank you for your response and yes, I know that now because of your reply but there's nothing on the cli help screen itself to indicate that `zellij [command] --help` even exists. Perhaps including `zellij [command] --help` on the default help screen would provide a better user experience and cut down on issues like this. BTW zellij is awesome and I promote it every chance I get. It is a terminal game changer in MHO.

**Comment by jeffschwartz** [comment]
Thanks @ottok

**Pull request closed without merging**
